pyop2/codegen/builder.py

```diff
@@ -442,6 +453,91 @@ def emit_unpack_instruction(self, *, loop_indices=None):
+        cshape = 0
+        # Need to compute row and col shape based on individual pack shapes
+        for p in self.packs[:, 0]:
+            shape, _ = p.shapes
+            rshape += numpy.prod(shape, dtype=int)
+        for p in self.packs[0, :]:
+            _, shape = p.shapes
+            cshape += numpy.prod(shape, dtype=int)
+        shape = (rshape, cshape)
+        if self.access in {WRITE, INC}:
+            val = Zero((), self.dtype)
+            multiindex = MultiIndex(*(Index(e) for e in shape))
+            pack = Materialise(PackInst(), val, multiindex)
+            self._pack = pack
+            return pack
+        else:
+            raise ValueError("Unexpected access type")
+
+    def kernel_arg(self, loop_indices=None):
+        pack = self.pack(loop_indices=loop_indices)
+        return Indexed(pack, tuple(Index(e) for e in pack.shape))
+
+    def emit_pack_instruction(self, *, loop_indices=None):
+        return ()
+
+    def emit_unpack_instruction(self, *,
+                                loop_indices=None):
+        pack = self.pack(loop_indices=loop_indices)
+        mixed_to_local = []
+        local_to_global = []
+        roffset = 0
+        for row in self.packs:
+            coffset = 0
+            for p in row:
+                rshape, cshape = p.shapes
+                pack_ = p.pack(loop_indices=loop_indices, only_declare=True)
+                rindices = tuple(Index(e) for e in rshape)
+                cindices = tuple(Index(e) for e in cshape)
+                indices = MultiIndex(*rindices, *cindices)
+                lvalue = Indexed(pack_, indices)
+                rextents = [numpy.prod(rshape[i+1:], dtype=numpy.int32) for i in range(len(rshape))]
+                cextents = [numpy.prod(cshape[i+1:], dtype=numpy.int32) for i in range(len(cshape))]
+                flat_row_index = reduce(Sum, [Product(i, Literal(IntType.type(e), casting=False))
+                                              for i, e in zip(rindices, rextents)],
+                                        Literal(IntType.type(0), casting=False))
+                flat_col_index = reduce(Sum, [Product(i, Literal(IntType.type(e), casting=False))
+                                              for i, e in zip(cindices, cextents)],
+                                        Literal(IntType.type(0), casting=False))
+
+                flat_index = MultiIndex(Sum(flat_row_index, Literal(IntType.type(roffset), casting=False)),
+                                        Sum(flat_col_index, Literal(IntType.type(coffset), casting=False)))
+                rvalue = Indexed(pack, flat_index)
+                # Copy from local mixed element tensor into non-mixed
+                mixed_to_local.append(Accumulate(PreUnpackInst(), lvalue, rvalue))
+                # And into global matrix.
+                local_to_global.extend(p.emit_unpack_instruction(loop_indices=loop_indices))
+                coffset += numpy.prod(cshape, dtype=numpy.int32)
```

Comment:
Okay, so the way this works is we run over the mixed matrix, extract the non-mixed matrices and invoke the distribution of the local contributions into the global matrix on each of them separately? In theory, could Slate do mixed matrixes built of matrices that are themselves mixed? If yes, does the unpacking work for them as well?

Reply:
So right now PyOP2 doesn't represent the concept of multiple levels of nesting (e.g. a mixeddat is always just a flat bag of dats). As a consequence the blocks in the matrix are always non-mixed, so this code works.